TESTS/mbed_hal/mpu/main.cpp

```diff
@@ -0,0 +1,151 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "utest/utest.h"
+#include "unity/unity.h"
+#include "greentea-client/test_env.h"
+
+#include "cmsis.h"
+#include <stdlib.h>
+
+#include "mpu_api.h"
+
+// Define included by mpu_api.h
+#if !MBED_MPU_ENABLED
+#error [NOT_SUPPORTED] MPU API not supported for this target
+#endif
+
+using namespace utest::v1;
+
+#define HARDFAULT_IRQn   ((IRQn_Type)-13)
+
+// Assembly return instruction: bx lr
+#define ASM_BX_LR 0x4770
+
+volatile uint32_t fault_count;
+uint32_t real_hard_fault_handler;
+
+static volatile uint16_t data_function = ASM_BX_LR;
+static volatile uint16_t bss_function;
+
+static void call_mem(const volatile uint16_t *mem_function)
+{
+    // or the address with 1 to ensure the thumb bit is set
+    ((void (*)()) ((uint32_t)mem_function | 1))();
+}
+
+static void hard_fault_handler_test()
+{
+    fault_count++;
+    mbed_mpu_enable_ram_xn(false);
```

Comment:
Do we need to turn of MPU here? We already do it as part of the test https://github.com/ARMmbed/mbed-os/pull/8335/files#diff-e91b21e2fbb82b6e0c8982f314540839R68

Reply:
Yeah, otherwise it will fault again when it returns from the hardfault.